gh-101100: Fix broken xrefs in fcntl module doc (#115691)
python/cpython: Doc/library/fcntl.rst

----------------

> This module performs file and I/O control on file descriptors.  It is
> an interface to the :c:func:`fcntl` and :c:func:`ioctl` Unix routines.

Reviewer: ```suggestion
This module performs file and I/O control on file descriptors. It is an
interface to the :c:func:`fcntl` and :c:func:`ioctl` Unix routines.
```

Minimize unnecessary diffs (see python/devguide#1018 )

----------------

This module performs file control and I/O control on file
descriptors. See the :manpage:`fcntl(2)` and :manpage:`ioctl(2)` Unix

> manual pages for full details.

Reviewer: If we're rewriting these lines anyways, seems better to not add new arbitrary line breaks that aren't motivated by either semantics or column width, and ensure we match the original more closely. Also, elide a redundant/repetitive word:

```suggestion
This module performs file and I/O control on file descriptors.
See the :manpage:`fcntl(2)` and :manpage:`ioctl(2)` Unix
manual pages for full details.
```

Also, to me, not being a low-level *nix expert, it seems there might be some value in being more explicit about the connection between this Python module and the *nix system man pages linked—the current wording seems to leave a bit of a gap up to the imagination of the reader, of which those with less experience may not be as comfortable implicitly filling in, and which is generally not ideal in a reference doc such as this. So perhaps:

```suggestion
This module performs file and I/O control on file descriptors.
See the :manpage:`fcntl(2)` and :manpage:`ioctl(2)` Unix manual pages
for full details on the interfaces this module wraps.
```

("Interfaces" could be replaced with "APIs", "functions", "library calls", etc. to preference)

Author: My take on this will be in the next push.

Reviewer: Sure, LGTM thanks; just a minor tweak to minimize an unnecessary reflow diff

----------------

This module performs file control and I/O control on file descriptors. It is an

> interface to the :c:func:`!fcntl` and :c:func:`!ioctl` Unix system calls.  For a

Reviewer 2: "system calls" is not accurate. These are POSIX C library `libc` functions. Routines was an appropriately ambiguous word that is basically shorthand for library calls. _(that they are system calls behind the scenes is an OS implementation detail)_

When updating text like this I tend to replace the Unix term with POSIX as that is actually relevant, it not being the '90s and all.

---

Why not shorten the text entirely and just make these be the :manpage: links directly instead of the last sentence in the paragraph existing at all?

Author: Hmmm... The manpage links are to section 2, which in my old Unix lizard brain reads as "system calls." Historically, I don't think that was a behind-the-scenes thing. There are differences between system calls and library routines (performance, task switch opportunity, etc). If that's changed over time for some Unix-y operating systems, it probably hasn't changed for all of them. (Are they still actual system calls on NetBSD, for example?)

I going to speak a quiet part loud which has been bugging me as I worked through other doc files. gh-101100 is (I think) meant to clean up broken references. The first few changes I made, I restricted my attention to that task. Then came a bunch of bike shedding about content. This time I decided I would fix some wording I thought was incorrect. Now I'm supposed to change this back. Can't win for losing.

   in the :mod:`fcntl` module, using the same names as used in the relevant C
   header files. The argument *arg* can either be an integer value, or a
   :class:`bytes` object. With an integer value, the return value of this

>    function is the integer return value of the C :c:func:`!fcntl` call.  When

Reviewer: As discussed, already silenced globally; no need for this here (same with the others)—greatly reduces the overall delta to fix this to just the more meaningful changes

   buffer 1024 bytes long which is then passed to :func:`ioctl` and copied back
   into the supplied buffer.

>    If the :c:func:`!ioctl` fails, an :exc:`OSError` exception is raised.

Reviewer: Same here, once you add it as discussed

   method are accepted as well) of the file to lock or unlock, and *cmd*
   is one of the following values:

   * :const:`!LOCK_UN` -- unlock
   * :const:`!LOCK_SH` -- acquire a shared lock

>    * :const:`!LOCK_EX` -- acquire an exclusive lock

Reviewer: These are valid, public constants of the `fcntl` module used in one if its primary public functions, so ISTM instead of silencing these valid warnings, these really should be properly documented, if even just by something simple like:

```rst
.. data:: LOCK_UN
          LOCK_SH
          LOCK_EX

   Flags accepted by :func:`lockf` to signal the operation to perform:
   UNlock, SHared-lock, and EXclusive-lock, respectively.
   See the :manpage:`fcntl(2)` manual page for more information.
```

Furthermore, it makes clear to the less experienced reader that these are module-level constants of the `fcntl` module, rather than strings or from somewhere else they missed; I'm not sure I as such would have necessarily been 100% clear on this otherwise without seeing the relevant explicit markup.

And this change would no longer be necessary:

```suggestion
   * :const:`LOCK_UN` -- unlock
   * :const:`LOCK_SH` -- acquire a shared lock
   * :const:`LOCK_EX` -- acquire an exclusive lock
```

Author: I'm inclined to leave them not described here. This is a low-level module which exposes `fcntl`, `ioctl` and `flock` system calls and their various constants. I think it best to refer the user to the C stdlib/system call documentation for the particulars.

If you want workable links I think some sort of `intersphinx` connection to manpages (like the `:manpage:` role, but more precise than an [entire URL](https://manpages.ubuntu.com/manpages/focal/en/man2/flock.2.html)) makes some sense, but that would appear to be a longer term project than just cleaning up a single module's documentation

Author: I decided to take a look at the flock(2) man page. One example is here:

https://man7.org/linux/man-pages/man2/flock.2.html

Note that the descriptions of these constants there are not really any more detailed than what is currently in the `fcntl` module doc. What's really needed is an index entry, not a reference to a nonexistent directive.

This seems to render reasonably, and seems to drop relevant entries in the global index:

```
   .. data:: LOCK_UN

             release an existing lock

   .. data:: LOCK_SH

             acquire a shared lock

   .. data:: LOCK_EX

             acquire an exclusive lock
```

Is there a more compact way to state that?

Reviewer: I really like the solution you came up with! It's a significant improvement over what I suggested as it avoids duplication and fragmentation and ensures the documentation is located right within the function for which it is relevant, while still generating the correct index entries and semantic markup (which I previously wasn't sure would work correctly inside a function, but I can confirm everything functions perfectly). And I think this could be a good fit a number of other places too that have similar module-global constants for specific functions that aren't formally documented.

> If you want workable links I think some sort of intersphinx connection to manpages (like the :manpage: role, but more precise than an [entire URL](https://manpages.ubuntu.com/manpages/focal/en/man2/flock.2.html)) makes some sense

For me I'm much more concerned with the index entries and semantic markup so users (and machines) can easily look them up (as it seems you are too, and your solution does perfectly well). However, thinking about it some more perhaps the simplest approach would be to extend the `manpage` directive to support linking to specific ids within the page, which the Debian manpages (which we currently use for `manpage) [fully support](https://manpages.debian.org/bookworm/manpages-dev/flock.2.en.html#LOCK_SH) down to a very fine granularity, including flags like these. However, as you say that's a non-trivial project and of course far out of scope for this PR.

> Is there a more compact way to state that?

Nope, unless you want to do like I had and put all three/four flags under the same directive (which means they all use the same description), but I greatly prefer what you have now and think its perfectly fine.

   When *cmd* is :const:`!LOCK_SH` or :const:`!LOCK_EX`, it can also be

>    bitwise ORed with :const:`!LOCK_NB` to avoid blocking on lock acquisition.

Reviewer: Oh wait, there's a `LOCK_NB` too? Seems to me that there's further chance here that the less experienced reader won't just assume that the newly introduced literal name happens to be a module-level constant of `fcntl` that has been left undocumented. Thus, likewise:

```rst
.. data:: LOCK_NB

   Flag that when bitwise-ORed with :const:`LOCK_SH` or :const:`LOCK_EX`
   signals :func:`lockf` not to block on lock acquisition.
```

And so, here:

```suggestion
   When *cmd* is :const:`LOCK_SH` or :const:`LOCK_EX`, it can also be
   bitwise ORed with :const:`LOCK_NB` to avoid blocking on lock acquisition.
```

Author: I reworked slightly. Please check the latest push.

Reviewer: Looks good! Just fixed a typo, added a tiny clarification and conformed the formatting per the style guide.

   If :const:`!LOCK_NB` is used and the lock cannot be acquired, an
   :exc:`OSError` will be raised and the exception will have an *errno*

>    attribute set to :data:`~errno.EACCES` or :data:`~errno.EAGAIN` (depending on the

Reviewer: ```suggestion
   attribute set to :const:`~errno.EACCES` or :const:`~errno.EAGAIN` (depending on the
```

`:const:` is equally syntactically valid and most semantically consistent and appropriate here, so no need to change the role type here.

Author: I changed it to `data` because `EAGAIN` and `EACESS` are defined in the `errno` module documentation using the `.. data::` directive.

Reviewer: The [`.. py:data::` directive](https://www.sphinx-doc.org/en/master/usage/domains/python.html#directive-py-data) is used to define both constants (`:const:`) and (possibly variable) module data (`:data:`), presumably since their definitions are syntactically identical in Python and differ only in their usage semantics. Therefore, the previous role was correct.

Also, in addition to the previously mentioned rationale (local consistency with other similar flags here, `const` being the most semantically appropriate choice as previously, and no reason to change the status quo), I also note that use of `:const:` instead of `:data:` to refer to `errno.E*` flags is standard convention across the docs, with 36 hits for ``:const:`[~!]{0,2}errno\.E`` and only 1 for ``:data:`[~!]{0,2}errno\.E``.

Author: reverted my change

   operating system; for portability, check for both values).  On at least some

>    systems, :const:`!LOCK_EX` can only be used if the file descriptor refers to a

Reviewer: ```suggestion
   systems, :const:`LOCK_EX` can only be used if the file descriptor refers to a
```

Likewise, this could and should be reverted

Author: Eh, providing a hyperlink to two or three lines above seems both cluttered and unnecessary. I think it should stay as-is. It would be different if `LOCK_EX` was defined elsewhere (like the `errno` constants).

(I was chided about something similar in earlier revisions of other library docs.)

Reviewer: Yeah, I might have been the one chiding you, haha. And indeed, with your change to locate the constant definition right above within this function, no need for this then :+1:

   method are accepted as well) of the file to lock or unlock, and *cmd*
   is one of the following values:

   .. data:: LOCK_UN

>              release an existing lock
> 
>    .. data:: LOCK_SH
> 
>              acquire a shared lock
> 
>    .. data:: LOCK_EX
> 
>              acquire an exclusive lock

Reviewer: ```suggestion
      Release an existing lock.

   .. data:: LOCK_SH

      Acquire a shared lock.

   .. data:: LOCK_EX

      Acquire an exclusive lock.
```

Fix formatting per [Style Guide](https://devguide.python.org/documentation/style-guide/#use-of-whitespace) and standard convention.

   .. data:: LOCK_NB

>              bitwise OR with any of the other three constantsoperations to make
>              the request non-blocking.

Reviewer: ```suggestion
      Bitwise OR with any of the other three ``LOCK_*`` constants to make
      the request non-blocking.
```

* Fix apparent typo
* Make it fully explicit which other constants are being referred to
* Fix formatting per Style Guide and standard convention, as above
